---
title: "A. grossa analysis"
author: "Edgar Caballero"
date: "`r Sys.Date()`"
output:
    html_document:
            keep_md: yes
            toc: true
            toc_depth: 5
            toc_float:
              collapsed: false
              smooth_scroll: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Stacks on a cluster

### Â¿ What's Stacks ?

Stacks is a bioinformatic software useful to analyze short sequence data such as Rad-seq for population genomic research. First, I will work with a dataset previously filtered, to build loci from short sequences, using the command `denovo_map.pl` in the cluster [Hydra](https://confluence.si.edu/display/HPC/High+Performance+Computing).

### `Denovo_map.pl` analysis 

First, we have to log in the cluster, and then place in the computer with the following code:

    pwd

Then we change to direction  in the cluster with the command `cd` :


    cd /scratch/genomics/caballeroeg /

Then, with the command `mkdir` we will create directory  to safe the files generated by  `denovo_map.pl` 

    mkdir -p optimization_denovo/m1 optimization_denovo/m2 optimization_denovo/m3 # and so it goes until we reach m9

We will create 9 folders inside the main "directory" to store the data for further analysis.

### Parameters optimization

`denovo_map.pl` is controled by three [main parameters](https://catchenlab.life.illinois.edu/stacks/param_tut.php) to generate the loci catalogue. The `-m` parameter specify the number of raw reads required to construct an initial stack (depth of coverage), `-M`controls the number of differences allowed between stacks to merge them as independent loci. Lastly, the `-n` parameter compare each loci, of the individuals with the catalogue to identify a loci as polymorphic or independent.

To set the best parameters combinations `-m`, `-M` y `-n` we have to create a subsample of our data, creating a `.txt file` with a part of our data called  `popt.txt`. I created this txt file by choosing randomly 5 individuals from each population with a coverage higher than 60 (>60) :


    1s02<tab>opt
    2s09F<tab>opt
    2s08<tab>opt
    2s25<tab>opt
    ...<tab>opt
    ...

It is important to separate the individuals and the population with  `tab`, otherwise the software will crash with an error. For the parameters test is not important to specify the population, purposely I used the 'opt' as population for all the individuals.

Once we defined our popmap file (popt.txt), the stacks files and the output director, we can execute `denovo_map.pl`, as a job-file:


``` /bin/sh
# ----------------Parameters---------------------- #
#$  -S /bin/sh
#$ -pe mthread 30
#$ -q mThM.q
#$ -l mres=240G,h_data=8G,h_vmem=8G,himem
#$ -cwd
#$ -j y
#$ -N o8
#$ -o o8.log
#$ -m bea
#$ -M caballeroeg@si.edu
#
# ----------------Modules------------------------- #
module load ~/modulefiles/miniconda
source activate ste
#
# ----------------Your Commands------------------- #
#
echo + `date` job $JOB_NAME started in $QUEUE with jobID=$JOB_ID on $HOSTNAME
echo + NSLOTS = $NSLOTS
#
denovo_map.pl -m 3 -M 8 -N 8 --samples /scratch/genomics/caballeroeg/CPrads/clean_rads/ --popmap /scratch/genomics/caballeroeg/optimization_denovo/m1/popt.txt  -o /scratch/genomics/caballeroeg/optimization_denovo/m8   --paired -r 0.8
 
#
echo = `date` job $JOB_NAME done

```
For each parameter test, we will vary the `-M` and `-n` from one to nine. For further details about the parameter ests click [here](https://www.biorxiv.org/content/biorxiv/early/2021/11/04/2021.11.02.466953.full.pdf) and visit the [Stacks website](https://catchenlab.life.illinois.edu/stacks/comp/denovo_map.php).

After finishing all the test, we can extract the quantity of loci generated by each `-M` and `-n` set, using the following command :

    $ cat populations.sumstats.tsv | \
     grep -v '^#' | \ # Remove the comments
     cut -f 1 | \ # Splits the first column (Locus ID)
     sort -n -u | \ #  Sort by numerical order and unique values
     wc -l # count the number of lines
     10065

For instance, for `-M`=1 and `n`=1 the number of polimorphic loci identified was 10065.We will do this for all the parameters and organize it in the following way

```{r}
optimization <- read.csv("r_80.csv",sep = ";")
optimization

```
